### ISSUE TITLE
Crash on entering the call

### ISSUE THREAD

**Issue by saif-o99** [react-native-webrtc/react-native-webrtc#1255]
#### Expected Behavior
this crash doesn't happen always, more likely to occur in video group calls

#### Observed Behavior
![MicrosoftTeams-image](https://user-images.githubusercontent.com/52207667/208480651-b6a17ef5-6f33-462b-a9bb-a70c7c9e0c23.png)

`*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***

 

Build fingerprint: 'google/sdk_gphone_x86/generic_x86_arm:11/RSR1.201013.001/6903271:user/release-keys'
Revision: '0'
ABI: 'x86'
Timestamp: 2022-12-19 19:46:40+0300
pid: 15274, tid: 15356, name: signaling_threa  >>> com.pingme2022 <<<
uid: 10159
signal 6 (SIGABRT), code -1 (SI_QUEUE), fault addr --------
    eax 00000000  ebx 00003baa  ecx 00003bfc  edx 00000006
    edi ed1cd81e  esi b2989c70
    ebp ee941b90  esp b2989c18  eip ee941b99
backtrace:
      #00 pc 00000b99  [vdso] (__kernel_vsyscall+9)
      #01 pc 0005ad68  /apex/com.android.runtime/lib/bionic/libc.so (syscall+40) (BuildId: 6e3a0180fa6637b68c0d181c343e6806)
      #02 pc 00076511  /apex/com.android.runtime/lib/bionic/libc.so (abort+209) (BuildId: 6e3a0180fa6637b68c0d181c343e6806)
      #03 pc 004c3c6c  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #04 pc 004c3f8f  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #05 pc 004a69a8  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #06 pc 00908714  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #07 pc 0090869b  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #08 pc 00477623  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #09 pc 00482c55  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #10 pc 0048712d  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #11 pc 004eee11  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #12 pc 004edf35  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #13 pc 004ef249  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #14 pc 004ef188  /data/app/~~UPQzE5ueBBIQir83PzrF5g==/com.pingme2022-OIGtfTSCXt8f5Ie85wfTbw==/base.apk!libjingle_peerconnection_so.so (offset 0x92cd000) (BuildId: 3ff11105582ab47f)
      #15 pc 000e6974  /apex/com.android.runtime/lib/bionic/libc.so (__pthread_start(void*)+100) (BuildId: 6e3a0180fa6637b68c0d181c343e6806)
      #16 pc 00078567  /apex/com.android.runtime/lib/bionic/libc.so (__start_thread+71) (BuildId: 6e3a0180fa6637b68c0d181c343e6806)`
#### Steps to reproduce the issue
enter a call

Platform Information
React Native Version: 0.70.6
Plugin Version: 106.0.1
OS: Android 12, 11, 9
OS Version: 9/12/10

**Comment by 8BallBomBom** [comment]
Bit of an unusual error.

Can see some other mentions of the same sort of error going back some years relating to starting and stopping streams at a fast rate which can cause exceptions and then crash the app.
Definitely shouldn't be happening though.

Also ProGuard should be configured for release versions.
```
-keep class com.oney.WebRTCModule.** { *; }
-keep class org.webrtc.** { *; }
```

**Comment by MusabBasheer** [comment]
```
12-14 23:49:55.102 23698 23947 D SoLoader: libnative-filters.so found on /data/app/~~hM8E-s2bZL55GEScWV3jWw==/com.pingme2022-0CJ2Tcks-5ksBZegBzJ7pg==/base.apk!/lib/x86
12-14 23:49:55.118 23698 23801 I ReactNativeJS: onnegotiationneeded
12-14 23:49:55.152 23698 23798 D com.oney.WebRTCModule.WebRTCModule: onAddTrack
12-14 23:49:55.153 23698 23798 D com.oney.WebRTCModule.WebRTCModule: onAddTrack
12-14 23:49:55.161 23698 23950 I org.webrtc.Logging: AndroidVideoDecoder: ctor name: OMX.android.goldfish.vp8.decoder type: VP8 color format: 19 context: org.webrtc.EglBase14Impl$Context@42d2054
12-14 23:49:55.169   321 22732 D goldfish-address-space: claimShared: Ask to claim region [0x3f188d000 0x3f1924000]
12-14 23:49:55.182 23698 23797 I org.webrtc.Logging: WebRtcAudioTrackExternal: initPlayout(sampleRate=44100, channels=1, bufferSizeFactor=1.0)
12-14 23:49:55.182 23698 23797 I org.webrtc.Logging: WebRtcAudioTrackExternal: byteBuffer.capacity: 882
12-14 23:49:55.186 23698 23797 I org.webrtc.Logging: WebRtcAudioTrackExternal: minBufferSizeInBytes: 3912
12-14 23:49:55.186 23698 23797 I org.webrtc.Logging: WebRtcAudioTrackExternal: createAudioTrackBeforeOreo
12-14 23:49:55.186 23698 23797 I org.webrtc.Logging: WebRtcAudioTrackExternal: nativeOutputSampleRate: 44100
12-14 23:49:55.190   368 17125 D AudioFlinger: Client defaulted notificationFrames to 978 for frameCount 1956
12-14 23:49:55.192   368 17125 D AF::TrackHandle: OpPlayAudio: track:180 usage:2 not muted
12-14 23:49:55.204 23698 23797 I org.webrtc.Logging: WebRtcAudioTrackExternal: AudioTrack: session ID: 1521, channels: 1, sample rate: 44100, max gain: 1.0
12-14 23:49:55.204 23698 23797 I org.webrtc.Logging: WebRtcAudioTrackExternal: AudioTrack: buffer size in frames: 1956
12-14 23:49:55.204 23698 23797 I org.webrtc.Logging: WebRtcAudioTrackExternal: AudioTrack: buffer capacity in frames: 1956
12-14 23:49:55.205 23698 23797 I org.webrtc.Logging: WebRtcAudioTrackExternal: startPlayout
12-14 23:49:55.213   512   661 I system_server: oneway function results will be dropped but finished with status OK and parcel size 4
12-14 23:49:55.219   321 22732 D goldfish-address-space: claimShared: Ask to claim region [0x3f1924000 0x3f19bb000]
12-14 23:49:55.221 23698 23955 I org.webrtc.Logging: WebRtcAudioTrackExternal: AudioTrackThread@[name=AudioTrackJavaThread, id=1115]
12-14 23:49:55.221 23698 23955 I org.webrtc.Logging: WebRtcAudioTrackExternal: doAudioTrackStateCallback: 0
12-14 23:49:55.247 23698 23956 I org.webrtc.Logging: AndroidVideoDecoder: ctor name: OMX.android.goldfish.vp8.decoder type: VP8 color format: 19 context: org.webrtc.EglBase14Impl$Context@42d2054
12-14 23:49:55.254 23698 23798 W System.err: java.lang.IllegalStateException: RtpTransceiver has been disposed.
12-14 23:49:55.254 23698 23798 W System.err:    at org.webrtc.RtpTransceiver.checkRtpTransceiverExists(RtpTransceiver.java:251)
12-14 23:49:55.254 23698 23798 W System.err:    at org.webrtc.RtpTransceiver.dispose(RtpTransceiver.java:242)
12-14 23:49:55.254 23698 23798 W System.err:    at org.webrtc.PeerConnection.getTransceivers(PeerConnection.java:1055)
12-14 23:49:55.254 23698 23798 W System.err:    at com.oney.WebRTCModule.WebRTCModule.getTransceiversInfo(WebRTCModule.java:861)
12-14 23:49:55.254 23698 23798 W System.err:    at com.oney.WebRTCModule.WebRTCModule.access$500(WebRTCModule.java:37)
12-14 23:49:55.254 23698 23798 W System.err:    at com.oney.WebRTCModule.WebRTCModule$3.onSetSuccess(WebRTCModule.java:1009)
12-14 23:49:55.254 23698 23798 E rtc     : #
12-14 23:49:55.254 23698 23798 E rtc     : # Fatal error in: gen/sdk/android/generated_metrics_jni/../../../../../../sdk/android/src/jni/jni_generator_helper.h, line 94
12-14 23:49:55.254 23698 23798 E rtc     : # last system error: 0
12-14 23:49:55.254 23698 23798 E rtc     : # Check failed: !env->ExceptionCheck()
12-14 23:49:55.254 23698 23798 E rtc     : # 
```

**Comment by saghul** [comment]
Hum a transceiver being disposed would imply the PC was closed I think. Is that the case?

**Comment by saghul** [comment]
What are you doing prior to `setLocalDescription` ? Here is what I see from that crash:

- sLD succeeds
- as part of the result we serialize the transceivers: https://github.com/react-native-webrtc/react-native-webrtc/blob/c20af102c1be44c90fc4415d2a844bada8a5cfe1/android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java#L1005
- first we get the native transceiver objects: https://github.com/react-native-webrtc/react-native-webrtc/blob/c20af102c1be44c90fc4415d2a844bada8a5cfe1/android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java#L857
- when doing that the Java PC wrapper will dispose the existing Java object references, before re-wrapping them from native: https://github.com/jitsi/webrtc/blob/464168b5e9db3a2ad9a4bfd160eb6d4792e3fd23/sdk/android/api/org/webrtc/PeerConnection.java#L1053
- turns out, one of those transceivers was already disposed! https://github.com/jitsi/webrtc/blob/464168b5e9db3a2ad9a4bfd160eb6d4792e3fd23/sdk/android/api/org/webrtc/RtpTransceiver.java#L242

We don't manage the lifetime of transceivers, so this feels like a WebRTC bug but we'd need to know what sequence of actions triggers it.

**Comment by saif-o99** [comment]
```
    onReceiveSdp: flow(function* ({ description, groupId, userId, glareNumber }): /* type annotation */ any {
      let polite = false
      let ignoreOffer = false

      try {
        if (description) {
          const connectionKey = self.getConnectionKey(groupId, userId)
          const connection: RTCPeerConnection = self.connections.get(connectionKey)
          if(!connection) return
          const readyForOffer =
            !self.connectionsData[connectionKey]?.makingOffer &&
            (connection.signalingState === "stable" || self.connectionsData[connectionKey]?.isSettingRemoteAnswerPending);
          const offerCollision = (description.type === "offer") && !readyForOffer
          polite = self.connectionsData[connectionKey]?.glareNumber < glareNumber
          ignoreOffer = !polite && offerCollision
          if (ignoreOffer) {
            return
          }

          if (description.type === "offer" && connection.signalingState !== "stable") {
            yield Promise.all([
              connection.setLocalDescription({ type: "rollback" } as any),
              connection.setRemoteDescription(description),
            ])
          } else {
            self.addConnectionData(connectionKey, { isSettingRemoteAnswerPending: description.type === "answer"});
            yield connection.setRemoteDescription(description)
            self.addConnectionData(connectionKey, { isSettingRemoteAnswerPending: false});
          }
          self.processCandidates(connectionKey)

          if (description.type === "offer") {
            yield connection.setLocalDescription()
            yield self.sendCallSdp(groupId, userId, connection.localDescription)
          }
        }
      } catch (err) {
        console.error(err)
      }

    }),
```

```
    createConnection: flow(function* (groupId: string, userId: string) {
      const iceServers = yield load("iceServers")
      const connection = new RTCPeerConnection({ iceServers: iceServers })
      const connectionKey = self.getConnectionKey(groupId, userId)
      const remoteStream = new MediaStream(undefined)
     // let makingOffer = false
      self.addConnectionData(connectionKey, { isSettingRemoteAnswerPending: false})
      connection.addEventListener("negotiationneeded", async () => {
        __DEV__ && console.log("onnegotiationneeded" ,userId)
        try {
          if ( connection.signalingState !== "stable" || (connection.localDescription && connection.iceConnectionState !== "failed")) return
          self.addConnectionData(connectionKey, { makingOffer: true, glareNumber: new Date().getTime() })
          // create offer
          await connection.setLocalDescription()

          await self.sendCallSdp(groupId, userId, connection.localDescription)
        }catch (e) {
          __DEV__ && console.log("error in negotiationneeded",e)
        } finally {
          self.addConnectionData(connectionKey, { makingOffer: false })
        }
      })
```

this is where i use the setLocalDescription. i'm not sure if i have to use transceivers or its internally thing but i don't use them in my code, i use add track

**Comment by saghul** [comment]
Any chance you can reproduce this with a smaller example?

**Comment by qusaieilouti99** [comment]
@saghul I'm working with him on the same example
It's a little hard to give you a smaller example but I can give you the steps for reproducing this crash.

I want t mention that we implemented the perfect negotiation flow after we had a glare happens sometimes so the solution was to follow a perfect negotiation approach.

1. we're creating the offers from both peers in the connection and sending a number with each offer.
2. when the other peer receives the offer, it will compare the received number with the local number he created (if the local number doesn't exist he assumes he is a polite peer )
3. and the rest as you know that the polite should roll back his offer and accept the remote offer then create the answer.

Now, where is the issue?
When one peer starts creating his offer and I receive the remote offer at the same time, then this if statement will be ignored:
```
if (description.type === "offer" && connection.signalingState !== "stable") {
            yield Promise.all([
              connection.setLocalDescription({ type: "rollback" } as any),
              connection.setRemoteDescription(description),
            ])
}
```
so the local offer is being created the status is still stable, so it assumes that I don't have a local offer to rollback since I actually do but it's not finished yet.
so it'll enter the else statement:
```
else {
       yield connection.setRemoteDescription(description)
 }
```

This is the exact flow when the crash happened, I hope this can help.

**Comment by qusaieilouti99** [comment]
Some logs for the crash:

```
1- 'onnegotiationneeded' at 11:55:29.396
2- 'create offer on negotiation' at 11:55:29.401
3- 'onReceiveSdp', 'offer' at 11:55:29.439
4- 'don't have local offer yet, so set the received SDP as remote description' at 11:55:29.439
5- ### the crash ### at 11:55:29.746
6- 'signalingstatechange', 'have-local-offer' at 11:55:30.070
```

You see the gap between (2) and (6) this is what cause the issue.

**Comment by saghul** [comment]
Hum, the crash should indeed not happen, but even if we fix it, I think there is a problem in your logic.

Creating an offer is an async operation, so you'll always run the risk of not catching it. You should probably save a flag indicating you are in the process of creating an offer, and clear it once the offer is ready.

So, summarizing, this happens when you have done done some addTrack, createOffer, then called setRemoteDescription. Is that correct?

**Comment by qusaieilouti99** [comment]
Yes, but exactly when the setRemoteDescription starts before creating the offer finishes

**Comment by saghul** [comment]
> Yes, but exactly when the setRemoteDescription starts before creating the offer finishes

Hum. I don't think that's it. All native calls are serialized, so even if sRD is called before cA resolves, internally, it will run after. The native call that is.

I suspect this may happen even if you don't call cA, just by having called addTrack and then do a sRD. I'm not 100% sure though.



> I noticed that I need that flag, but I'm wondering why all the examples out there don't mention these small race conditions Like, am I the only one who faces these issues, or do the examples just mention the happy day scenario? @saghul

Isn't that because we don't support one part of the auto-rollback? Memory is fuzzy now, but IIRC there is something about perfect negotiation we don't yet support isn't there?

**Comment by qusaieilouti99** [comment]
> Isn't that because we don't support one part of the auto-rollback? Memory is fuzzy now, but IIRC there is something about perfect negotiation we don't yet support isn't there?

yeah, this might be the reason, but I'm in shock actually after trying to remove the manual rollback logic and the flow is still working fine, did you add it in the last beta releases?

**Comment by qusaieilouti99** [comment]
What I mean is I do:
1- setLocalDescription() => (create offer automatically)
2- setRemoteDescription(offer) => setting the remote offer as RemoteDescription
3- setLocalDescription() => (create answer automatically) 

This works fine !!

**Comment by saghul** [comment]
> > Isn't that because we don't support one part of the auto-rollback? Memory is fuzzy now, but IIRC there is something about perfect negotiation we don't yet support isn't there?
> 
> yeah, this might be the reason, but I'm in shock actually after trying to remove the manual rollback logic and the flow is still working fine, did you add it in the last beta releases?

We added nothing specific to perfect negotiation but there is a non-zero chance it might have been fixed by accident :-)

**Comment by qusaieilouti99** [comment]
I think it's fixed because I tested it many times and it rolls it back automatically. now I'll test if this affects the crash would be great.

**Comment by qusaieilouti99** [comment]
@saghul In the current state, the crash is almost gone.
I faced it again only once but wasn't logging anything at that point so I can't tell the scenario.
I don't know if you can fix the crash internally or you still don't have enough information.

**Comment by qusaieilouti99** [comment]
The changes that happened to the logic:
```
1- I'm using now the auto rollBack instead of the manual.
2- while I'm creating an offer, I don't process any remote SDP I receive at that point, instead,
 I save it somewhere then after the offer is created successfully I process the remote SDP.
```

point (2) is the case that was crashing, so it's like a workaround and I think we should prevent the crash internally

**Comment by qusaieilouti99** [comment]
I have successfully reproduced the crash again. but at this time it's related to something else.

`java.lang.NullPointerException: Attempt to read from field 'org.webrtc.SessionDescription$Type org.webrtc.SessionDescription.type' on a null object reference at com.oney.WebRTCModule.PeerConnectionObserver.onIceGatheringChange(PeerConnectionObserver.java:441)`

Crash steps:
1- I created an offer 
2- I received a remote offer 
3- the setRemoteDescription rolled back the local description 
4- I received com.oney.WebRTCModule.WebRTCModule : onIceGatheringChangeCOMPLETE
5- inside onIceGatheringChange :  
```
SessionDescription newSdp = peerConnection.getLocalDescription();
            WritableMap newSdpMap = Arguments.createMap();
            newSdpMap.putString("type", newSdp.type.canonicalForm());
            newSdpMap.putString("sdp", newSdp.description);
            params.putMap("sdp", newSdpMap);
```
here the local description has been rolled back so it crashed because we can't call ".type" on null value which is the local description.

**Comment by saghul** [comment]
Yep, good catch. We'll get that fixed after the break ;-)

If you want to give it a go and send a PR that would be welcome too.

**Comment by qusaieilouti99** [comment]
#1262 fixed @saghul

**Comment by 8BallBomBom** [comment]
@qusaieilouti99 I've pulled your changes into #1268 along with the same fixes for iOS, any chance you can test that everything is working as expected on iOS?

We can run some general testing but it would be great to ensure everything is working properly on your end.

**Comment by saghul** [comment]
Merged #1268 I'll be cutting a release soon.
